lib/engine/src/artifact.rs

```diff
@@ -71,6 +71,11 @@ pub trait Artifact: Send + Sync {
         Ok(())
     }
 
+    /// Create a header file for use with a native object file;
+    fn create_header_file(&self) -> Option<String> {
```

Comment:
I'm not sure that we should modify the engine. It seems this case is only needed for this specific engine?

Reply:
Yeah the issue is that it needs access to the internals.

The way I see it:
- We can add trait methods that allow us to downcast and get specific implementations out of `dyn Artifact`
- We can have a method like `create_header_file`
- We can attempt to change the calling code so we don't deal with abstract types, we deal with `ObjectFileEngine` directly
- We can restructure the way symbol naming in object files work and separate the logic of header file generation from this engine
- We can expose an API for symbol naming via `Artifact` so the header file generation code can get the information it needs